Service/Product/FacetBuilder.php

```diff
@@ -0,0 +1,242 @@
+use Magento\Store\Model\StoreManagerInterface;
+
+class FacetBuilder
+{
+    public const FACET_ATTRIBUTE_PRICE = 'price';
+    public const FACET_ATTRIBUTE_CATEGORIES = 'categories';
+    public const FACET_ATTRIBUTES_CATEGORY_ID = 'categoryIds';
+
+    public const FACET_KEY_ATTRIBUTE_NAME = 'attribute';
+    public const FACET_KEY_SEARCHABLE = 'searchable';
+
+    public const FACET_SEARCHABLE_SEARCHABLE = '1';
+    public const FACET_SEARCHABLE_NOT_SEARCHABLE = '2';
+    public const FACET_SEARCHABLE_FILTER_ONLY = '3';
+
+    // Local raw facet cache, indexed by $storeId
+    protected array $facets = [];
+
+    public function __construct(
+        protected ConfigHelper                            $configHelper,
+        protected StoreManagerInterface                   $storeManager,
+        protected CurrencyHelper                          $currencyManager,
+        protected GroupCollection                         $groupCollection,
+        protected GroupExcludedWebsiteRepositoryInterface $groupExcludedWebsiteRepository,
+    )
+    {}
+
+    /**
+     * @param int $storeId
+     * @return string[]
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    public function getAttributesForFaceting(int $storeId): array
+    {
+        return array_map(
+            function($facet) {
+                return $this->decorateAttributeForFaceting($facet);
+            },
+            $this->getRawFacets($storeId)
+        );
+    }
+
+    /**
+     * @param int $storeId
+     * @return array<string, array>|null
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    public function getRenderingContent(int $storeId): ?array
+    {
+        $facets = $this->getRawFacets($storeId);
+        if (empty($facets)) {
+            return null;
+        }
+        $attributes = $this->extractFacetAttributeNames($facets);
+        return [
+            'facetOrdering' => [
+                'facets' => [
+                    'order' => $attributes
+                ],
+                'values' => $this->getRenderingContentValues($attributes)
+            ],
+        ];
+    }
+
+    /**
+     * @param array<array<string, mixed>> $facets
+     * @return string[]
+     */
+    protected function extractFacetAttributeNames(array $facets): array
+    {
+        return array_map(
+            function($facet) {
+                return $facet[self::FACET_KEY_ATTRIBUTE_NAME];
+            },
+            $facets
+        );
+    }
+
+    /**
+     * @param string[] $facets
+     * @return array<string, array>
+     */
+    protected function getRenderingContentValues(array $facets): array
+    {
+        return array_combine(
+            array_map(
+                function(string $attribute) {
+                    if ($attribute === self::FACET_ATTRIBUTE_CATEGORIES) {
+                        $attribute = self::FACET_ATTRIBUTE_CATEGORIES . '.level0';
```

Comment:
Is it normal that I don't get the `.level0` in my Algolia dashboard and that the `sortRemaingBy` attribute is not applied ? 

<img width="830" alt="image" src="https://github.com/user-attachments/assets/d3baa5dc-9474-496d-8344-df0a78b7cca2" />

Reply:
Fixed:
<img width="1111" alt="2025-03-05_08-51-01" src="https://github.com/user-attachments/assets/f5193a88-73c9-4bf3-bbfc-7e8eb7e935e2" />
Thanks for catching this!

Will add unit test later.

Reply:
Works on my side as well now :)

Reply:
Unit test added
<img width="1488" alt="image" src="https://github.com/user-attachments/assets/bd4fcdba-550e-46c4-920d-bf5ac82413f9" />